hikari/interactions/modal_interactions.py

```diff
@@ -0,0 +1,343 @@
+# -*- coding: utf-8 -*-
+# cython: language_level=3
+# Copyright (c) 2020 Nekokatt
+# Copyright (c) 2021 davfsa
+#
+# Permission is hereby granted, free of charge, to any person obtaining a copy
+# of this software and associated documentation files (the "Software"), to deal
+# in the Software without restriction, including without limitation the rights
+# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in all
+# copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+"""Models and enums used for Discord's Modals interaction flow."""
+
+from __future__ import annotations
+
+__all__: typing.List[str] = [
+    "ModalResponseTypesT",
+    "ModalInteraction",
+    "PartialTextInput",
+    "ModalInteraction",
+]
+
+import typing
+
+import attr
+
+from hikari import channels
+from hikari import guilds
+from hikari import messages
+from hikari import snowflakes
+from hikari import traits
+from hikari.interactions import base_interactions
+from hikari.internal import attr_extensions
+
+if typing.TYPE_CHECKING:
+    from hikari import users as _users
+    from hikari.api import special_endpoints
+
+ModalResponseTypesT = typing.Literal[
+    base_interactions.ResponseType.MESSAGE_CREATE, 4, base_interactions.ResponseType.DEFERRED_MESSAGE_CREATE, 5
+]
+"""Type-hint of the response types which are valid for a modal interaction.
+
+The following types are valid for this:
+
+* `hikari.interactions.base_interactions.ResponseType.MESSAGE_CREATE`/`4`
+* `hikari.interactions.base_interactions.ResponseType.DEFERRED_MESSAGE_CREATE`/`5`
+"""
+
+
+@attr.define(kw_only=True, weakref_slot=False)
+class PartialTextInput(messages.PartialComponent):
+    """A text input component in a modal interaction."""
+
+    custom_id: str = attr.field(repr=True)
+    """Developer set custom ID used for identifying interactions with this modal."""
+
+    value: str = attr.field(repr=True)
+    """Value provided for this text input."""
+
+
+class ModalInteractionActionRow(typing.Protocol):
```

Comment:
Using a protocol here kinda feels like an anti-pattern, this could easily be forgotten about while adding message components and since you cast to a protocol there'll be no type assertions flagging that it needs to be updated. Have you considered using a separate dataclass for this since its such a simple dataclass and while overlap between message and text input is unlikely this can always be handled down the line. 

How much was keeping the message component and modal component types and enums separate considered btw? Rn this seems to kind of infer that modal components are message components based on the type enum (which isn't really the case and seems more like copying over discord's fudged semantics from the docs) and while this asserts that only PartialTextInput will ever be set for modal components, there doesn't seem to be any relevant change which shows that PartialTextInput will never be set on messages (for example would it be worth it also switching over to usign a union of types on message objects rather than typing it as PartialComponent or just using a different base class for Modal components?)

Reply:
I'd like to avoid separating enums but moving `TextInputStyle` to a different module might be a good step towards separating different kinds of components.

Unions instead of `PartialComponent` sound good but I feel it'd be inconsistent with the rest of the library. Unless y'all plan on using unions elsewhere in the future in which case I'm all for it.

Reply:
I suppose in the meantime while this is being decided I'll get rid of `ModalInteractionActionRow` and it can be substituted for something else later.

Reply:
Let's just say this is outside the scope of this issue for now.

Reply:
Honestly the model component classes should probably just use a separate base class imo, they're just a separate thing from Message components with different semantics and translating Discord's docs being poorly structured over to here feels like a bad idea.

Reply:
Does this still hold true with the (upcoming) inclusion of selects in modals?

Reply:
It has been done